### DISCUSSION TITLE
Hope to discuss about PIV smart card

### DISCUSSION THREAD

**Discussion by crazydev0206** [OpenSC/OpenSC#3204] [Ideas]
Hope to discuss the implementation of read/write functions for PIV smart cards in the OpenSC project.

**Comment by dengert** [comment]
As I replied to someone else yesterday:

> NIST defines the PIV card, others like Yubikey have implemented parts of it, but NIST leaves it up
> to the card vendors to provide provisioning. NIST does define the 9B key, but does not require its
> use in provisioning and do define how to generate a key on card, but not how to import a key to the card.
> 
> The PIV was expected to be provisioned by a card management system and  not by the end user.
> The CMS would also take care of a lot of the details.
> 
> OpenSC implements most of 800-73-4:
>  https://csrc.nist.gov/pubs/sp/800/73/4/upd1/final
> 
> I have not looked at 800-73-5 as it is very new -  July 15, 2024  - 4 days ago!
> 
> The OpenSC piv-tool was designed to be be used with test cards, not production and uses just what
> is define to generate a key on card and load files and change pins. tools/piv-tool.c calls
> libopensc/card-piv.c to do the APDUs
> 
> So the OpenSC minidriver can be used with PIV card, but the registry entries are not installed be default
> because every PIV vendor has their own minidriver, that may include provisioning code and be installed
> via PnP on windows.

Also see: #3159 and  #3167 
 
 I  Also also replied with a second e-mail:
 
> The commands in NIST sp800-73 are based on a subset of ISO 7816-4 For example
> the GET_DATA and PUT_DATA referrer to objects not files.
> 
> OpenSC was originally written to support smart cards that support a PKCS15 file structure.
> The libopensc/card-<card>.c are card drivers that handle the ISO7816-4 commands.
> The libopensc/pkcs15-<card>.c handle special cases of PKCS15 data or commands or
> emulate PKCS15 functions that call the card driver as needed.
> 
> card-piv.c and pkcs15-piv.c are such files.
> 
> PIV does not use a PKCS15 file system. The "objects" on a PIV card are defined in 800-73.
> 
> OpenSC also implements PKCS11 that calls PKCS15 routines, and the OpenSC minidriver
> also calls pkcs15 routines.
> 
> The pkcs15init directory contains  profiles for pkcs15 type cards and some others but not
> all cards, and card provisioning was only done via one or more of the tools. It is only in the last
> few years, has code been added to PKCS11 to call the functions in the pkcs15init directory.
> 
> The tools directory has a number of tools that work at different levels:
> opensc-tool.c has the -s option, where one can provide an ISO7816-4 APDU to be issued to a card.
> 
> opensc-explorer.c deals with ISO 7816-4 files and objects.
> 
> pkcs15-tool and pkcs15-init deals with PKC15 structures.
> 
> pkcs11-tool.c is a general tool that calls PKCS11 modules.
> 
> Several other tools such as piv-tool can call PKCS15 routines ot the card driver directly.
> 
> As I said before NIST did not define how to initialize and provision PIV cards, leaving that up to vendors,
> who can use what ever they want. Some may use only what is defined in 800-73, others
> may use SCP01 or SCP03, or other methods.
> 
> Since every PIV card vendor provides there own minidriver, which does not need to use OpenSC,
> it is out of scope to try and add vendor specific code to OpenSC for provisioning.

**Comment by dengert** [comment]
I should point out, that there was one change added to OpenSC
https://github.com/OpenSC/OpenSC/pull/2671

I don't know how they are using this, but it looks like this defines access to any
PKCS15 underlying file structure that may be present in a card with their PIV applet.
From the comments I assume their code can use this from their minidriver.

**Reply by frankmorgner** [reply]
They are using it with a proprietary driver for pkcs15init. They needed #2671 so that their initialization can be triggered correctly, but the whole initialization process itself is not part of OpenSC.
